p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,426 @@
+package holepunch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p/p2p/protocol/holepunch/pb"
+	"github.com/libp2p/go-libp2p/p2p/protocol/identify"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-msgio/protoio"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+// TODO Should we have options for these ?
+var (
+	// Protocol is the libp2p protocol for Hole Punching.
+	Protocol protocol.ID = "/libp2p/holepunch/1.0.0"
+	// HolePunchTimeout is the timeout for the hole punch protocol stream.
+	HolePunchTimeout = 1 * time.Minute
+
+	maxMsgSize  = 4 * 1024 // 4K
+	dialTimeout = 5 * time.Second
+	maxRetries  = 5
+	retryWait   = 2 * time.Second
+)
+
+var (
+	log = logging.Logger("p2p-holepunch")
+)
+
+// The HolePunchService is used to make direct connections with a peer via hole-punching.
+type HolePunchService struct {
```

Comment:
This should probably just be named `Service`, since the package name is already `holepunch`.